sqlalchemy-stubs/sql/type_api.pyi

```diff
@@ -32,87 +63,85 @@ class TypeEngine(Traversible, Generic[_T]):
         def reverse_operate(self, op: Any, other: Any, **kwargs: Any): ...
         def __reduce__(self): ...
     hashable: bool = ...
-    comparator_factory: Any = ...
-    sort_key_function: Any = ...
+    comparator_factory: Type[Any] = ...
+    sort_key_function: Optional[_SortKeyFunction] = ...
     should_evaluate_none: bool = ...
-    def evaluates_none(self): ...
-    def copy(self, **kw: Any): ...
+    def evaluates_none(self: _TE) -> _TE: ...
+    def copy(self: _TE, **kw: Any) -> _TE: ...
     def compare_against_backend(
         self, dialect: Any, conn_type: Any
-    ) -> None: ...
-    def copy_value(self, value: Any): ...
-    def literal_processor(self, dialect: Any) -> None: ...
-    def bind_processor(self, dialect: Any) -> None: ...
-    def result_processor(self, dialect: Any, coltype: Any) -> None: ...
-    def column_expression(self, colexpr: Any) -> None: ...
-    def bind_expression(self, bindvalue: Any) -> None: ...
-    def compare_values(self, x: Any, y: Any): ...
-    def get_dbapi_type(self, dbapi: Any) -> None: ...
+    ) -> Any: ...
+    def copy_value(self, value: _T) -> _T: ...
+    def literal_processor(self, dialect: Any) -> Optional[_LiteralProcessor[_T]]: ...
+    def bind_processor(self, dialect: Any) -> Optional[_BindProcessor[_T]]: ...
+    def result_processor(self, dialect: Any, coltype: Any) -> Optional[_ResultProcessor[_T]]: ...
+    def column_expression(self, colexpr: Any) -> Any: ...
+    def bind_expression(self, bindvalue: Any) -> Any: ...
+    def compare_values(self, x: Any, y: Any) -> bool: ...
+    def get_dbapi_type(self, dbapi: Any) -> Any: ...
     @property
     def python_type(self) -> Type[_T]: ...
-    def with_variant(self, type_: Any, dialect_name: Any): ...
-    def as_generic(self, allow_nulltype: bool = ...): ...
-    def dialect_impl(self, dialect: Any): ...
-    def adapt(self, cls: Any, **kw: Any): ...
-    def coerce_compared_value(self, op: Any, value: Any): ...
-    def compile(self, dialect: Optional[Any] = ...): ...
+    def with_variant(self, type_: Type[TypeEngine[_U]], dialect_name: str) -> Variant[_U]: ...
```

Comment:
_U is to disambiguate from _T ?   if this were "_T" would that imply mypy would always expect "_T" thats coming into the method to be the "_T" of the type overall?

Reply:
Yes and yes